[NVPTX] Add errors for incorrect CUDA addrpaces

LewisCrawford opened the pull request:
The CUDA API only accepts kernel params in the global and generic address spaces, so display an error message when attempting to emit pointers outside those address-spaces from CUDA (but still allow them for OpenCL).

LewisCrawford committed:
[NVPTX] Add errors for incorrect CUDA addrpaces

The CUDA API only accepts kernel params in the global and generic
address spaces, so display an error message when attempting to
emit pointers outside those address-spaces from CUDA (but still
allow them for OpenCL).

gonzalobg reviewed (approved):
LGTM, thank you Lewis.

AlexMaclean commented:
One thing I wonder is whether this sort of logic really belongs in assembly printing. It seems like this verification could happen at any point during compilation, such as during an IR pass. Are there examples from other targets of target-specific IR verification? Do you have any thoughts about alternative locations for this verification?

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:
```diff
@@ -1399,19 +1399,27 @@ void NVPTXAsmPrinter::emitFunctionParamList(const Function *F, raw_ostream &O) {
           O << " .shared";
           break;
         case ADDRESS_SPACE_CONST:
+          if (IsCUDA)
+            report_fatal_error(".const ptr kernel args unsupported in CUDA.");
```
While local/shared pointers have no meaning outside of thread/CTA. `.const` pointers do exist and can be passed around.

gonzalobg replied:
For kernel parameters?

Artem-B replied:
On a second thought, I'm not quite convinced that we want this change.

> The CUDA API only accepts kernel params in the global and generic address spaces, so display an error message when attempting to emit pointers outside those address-spaces from CUDA (but still allow them for OpenCL).

@LewisCrawford can you elaborate on the motivation and the use case for this change?

CUDA operates on *generic* pointers, as far as the front-end is concerned, so I'm not quite sure which CUDA APIs you have in mind here that may be affected if we generate `.ptr .const`? 

Can you give me a few examples where these errors should be triggered when we compile CUDA sources? The included tests all operate on the IR level, so I can't tell what is the problem that this patch wants to address.

On the IR level, erroring out does not make whole lot of sense to me.  If the compilation of the CUDA sources can lead to such an IR, it should be diagnosed by the front-end. If we made this far in the compilation pipeline, I do not see why we would want to error out here, attempting to generate what appears to be a perfectly valid PTX.
What can possibly go wrong here that it's LLVM's job to prevent?

gonzalobg replied:
IIRC this change was motivated as a quality of life improvement for developers of non-cuda frontends (openmp/-acc offloading, c++ parallel algorithms, ...) that support a variety of backends (cuda ptx, opencl, ...). The set of state spaces each such backend supports for pointers passed as kernel arguments differs, leading to bugs in those frontends. IIRC, not all ptxas versions diagnose these sort of issues correctly, making the life of those frontend developers unnecessarily miserable. The LLVM NVPTX backend does support a wide range of ptxas versions, so detecting these issues there and providing a clear and concise error message about what went wrong seemed valuable.

LewisCrawford replied:
I've updated the patch to fix the `CodeGenCUDA/memcpy-libcall.cu` test, which shows that even the clang frontend with a `-cuda` triple can still generate invalid PTX which will crash when executed in a CUDA context. This will cause a confusing `CUDA_ERROR_INVALID_IMAGE` error message at run-time (similar to the one experienced here https://github.com/llvm/llvm-project/pull/114874#issuecomment-2658056661 ), rather than failing to compile and giving the more explicit error message I've introduced in this patch.

Unfortunately, this issue about CUDA only supporting .global and .generic pointers in kernel args, is one of several cases that are missing in the PTX documentation, where the PTX is technically legal according to the spec, but is invalid when used within the CUDA API (some texture-related queries is another case of this incompatibilty). We plan on adding public documentation about this soon though.

In the mean-time, it seems better to me to have some kind of readable error message about this to help implementers of various front-ends, even if the docs are currently missing it, since the `CUDA_ERROR_INVALID_IMAGE` caused when running this does not provide any useful information about the problem.

Artem-B replied:
> shows that even the clang frontend with a -cuda triple can still generate invalid PTX which will crash when executed in a CUDA context.

When users use `__attribute__((address_space(5)))` incorrectly/inappropriately, it's on them. There are infinitely many ways to write compileable, but nonfunctional code, and CUDA is not an exception.

I'd like  to understand what exactly the problem is with those wrong-AS pointers.
Is it the argument declaration as a the pointer  in non-generic/global AS that causes the runtime failure?
Or is that dereferencing of that pointer?

If it's the dereferencing that causes trouble, then there may be an argument that users should be allowed to pass around wrong-AS pointer arguments without dereferencing, if they want to. Sort of like C++ allows passing around `nullptr` -- we do know that we can't dereference it, but passing around the pointer value itself is fine. It's just a bunch of bits.

> PTX is technically legal according to the spec, but is invalid when used within the CUDA API [...]. We plan on adding public documentation about this soon though.

Awesome. If it is something that's going to be part of PTX spec, then enforcing it in LLVM may be OK, but it would be great to see the details first. 

There's also a question of *where* is the right place to enforce this restriction. Asm printer is understandably convenient, but it's far too late in the compilation pipeline and is removed from the user-supplied IR. Ideally we want to associate the error with specific part of user input (and LLVM is not great at diagnostics to start with).

I wonder if it should be done somewhere in the input IR validation where we can still dump the problematic IR.

LewisCrawford replied:
>Is it the argument declaration as a the pointer in non-generic/global AS that causes the runtime failure?
Or is that dereferencing of that pointer?

The example in https://github.com/llvm/llvm-project/pull/114874#issuecomment-2658056661 is an empty program that does not dereference anything. Merely launching a kernel with a pointer declared in those address-spaces is enough to trigger a run-time error, even if the pointer declarations are unused.

I agree that Asm printer is a convenient but suboptimal place for this. I'll look in to moving this to an IR validation step somewhere earlier in the pipeline.

LewisCrawford replied:
Moved to IR Verifier

LewisCrawford committed:
Update clang CodeGenCUDA/memcpy-libcall.cu test

Fix a clang test which failed with these new error messages,
as it illegally passed in kernel args in the .local address
space while using the -cuda triple.

LewisCrawford committed:
Fix NVPTX/lower-args-nvcl.ll test

Update the lower-args-nvcl.ll test to use the newly preferred
untyped load/stores to b64 instead of the u64 ones the origninal
version of the test used.

LewisCrawford committed:
Move checks int IR Verifier

Move the CUDA address-space checks out of ASMPrinter
and into the IR Verifier.

Artem-B commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -2927,6 +2928,20 @@ void Verifier::visitFunction(const Function &F) {
+      for (const Argument &Arg : F.args()) {
+        if (Arg.getType()->isPointerTy()) {
+          auto AS = Arg.getType()->getPointerAddressSpace();
+          Check(AS != NVPTXAS::AddressSpace::ADDRESS_SPACE_SHARED,
+                ".shared ptr kernel args unsupported in CUDA.", &Arg, &F);
```
I think the check should be rephrased to only allow generic and grlobal AS, and error out on anything else, so we don't have to update it when a new AS is added, or if/when someone uses a nonsensical AS.
